nep-monitor.mediawiki

```diff
@@ -0,0 +1,171 @@
+
+==Motivation==
+
+There is currently insufficient insight into NEO’s network quality and performance in part due to lack of data collection capabilities and available data attributes to collect. Network insight is important for 
+* Developers & investors assessing network suitability for their business case
+* Developers looking to improve network performance
+* Token holders looking to make an educated vote for consensus node candidates
+* (Future) Token holders to build or maintain trust in the network
+* Node hosters looking for hosting requirements and monitoring tools
+* Media outlets comparing blockchain platforms
+
+Current data attribute collection happens for most part via RPC calls which requires RPC server availability (enabled and port exposed). RPC calls only allow for collecting a single data attribute per call therefore requiring multiple RPC calls to collect all attributes. This pull based mechanism will put unnecessary load on the node when collecting data and repeats for every client requesting this data. If RPC server capabilities are not supported in the particular client, enabled in the configuration, or ports are not properly forwarded when supported, valuable node statistics are lost. Therefore, we feel a new monitoring interface and data attribute set that keeps the node load to a minimum while expanding insights is justified.
+
+
+==Specification==
+
+The specification consists of three parts 
+# architectural changes for exposing the monitor interface in a load efficient manner 
+# a description of the <code>monitor</code> payload format 
+# a description of the new data attribute set.
+
+===Monitor interface===
+The monitor interface is the access point where collection clients can access the data attributes. When a client connects to a node over the P2P or Websocket port a handshake protocol is started. First, the node sends a [http://docs.neo.org/en-us/network/network-protocol.html#network-message network message] with a <code>version</code> command and the appropriate payload. In the original handshake implementation the client is expected to respond with a similar <code>version</code> command network message and its own payload or be disconnected from the node. According to this proposal, we allow the client to provide an alternative response in the form of a network message holding the <code>monitor</code> command type and a zero length payload to indicate that the client is not a regular peer but a client that wishes to collect data attributes. The node separates the connection into its own monitoring channel on which no [http://docs.neo.org/en-us/network/network-protocol.html#network-message regular network messages] from the client are accepted and only data from the node is broadcasted. The <code>monitor</code> payload format is described in the next section. 
```

Comment:
Maybe hidden agents monitoring the network could be enough. Plugins developed for different purposes could handle this task.